File: AlgorithmsAndDataStructures/mod2/Heap.py
class Heap:
    def __init__(self):
        self.items = dict()     # key - (value, index)
        self.indexes = []       # index  - key // to know indexes 


    # Usefull functions
    def swap(self, i, j):
        x = self.indexes[i]     # key of 1 item
        y = self.indexes[j]     # key of 2 item

        # swap keys in index array
        self.indexes[i] = y
        self.indexes[j] = x

        temp = self.items[x][1]     # index of 1 item

        # swap indexes in dictionary
        self.items.update({x: (self.items[x][0], self.items[y][1])})
        self.items.update({y: (self.items[y][0], temp)})

    def bigger(self, i, j):
        if self.indexes[i] <= self.indexes[j]:
            return False
        else:
            return True


    # Check family UwU
    def hasParent(self, i):
        if (i - 1)/2 >= 0:
            return True
        return False
    def parentIndex(self, i):
        return int((i - 1)/2)

    def hasLeft(self, i):
        if i*2 + 1 < len(self.indexes):
            return True
        return False
    def leftIndex(self, i):
        return int(i*2 + 1)

    def hasRight(self, i):
        if i*2 + 2 < len(self.indexes):
            return True
        return False
    def rightIndex(self, i):
        return int(i*2 + 2)


    # heapifys
    def heapifyUp(self, i=None):
        if i:
            index = i
        else:
            index = len(self.indexes) - 1

        while self.hasParent(index) and self.bigger(self.parentIndex(index), index):
            self.swap(self.parentIndex(index), index)
            index = self.parentIndex(index)

    def heapifyDown(self, i=0):
        index = i
        while self.hasLeft(index):
            smaller = self.leftIndex(index)
            if self.hasRight(index) and self.bigger(self.leftIndex(index), self.rightIndex(index)):
                smaller = self.rightIndex(index)

            if self.bigger(smaller, index):
                break
            else:
                self.swap(index, smaller)

            index = smaller


    # all needed methods
    def add(self, key, data):
        if self.items.get(key, None):
            raise(Exception)

        self.items[key] = (data, int(len(self.indexes)))
        self.indexes.append(key)
        self.heapifyUp()

    def set(self, key, data):
        temp = self.items.get(key, None)
        if not temp:
            raise(Exception)

        self.items[key] = (data, temp[1])

    def delete(self, key):
        temp = self.items.get(key, None)
        if not temp:
            raise(Exception)

        if len(self.indexes) > 1:
            lastKey = self.indexes[-1]
            last = self.items.get(lastKey, None)

            # set last item index of deleted
            self.items.update({lastKey: (last[0], temp[1])})

            # set key of last item to deleted index
            self.indexes[temp[1]] = lastKey

        self.indexes.pop()
        del self.items[key]

        if temp[1] < len(self.indexes):   # dont heapify if deleted last element
            self.heapifyDown(i=temp[1])
            self.heapifyUp(i=temp[1])

    def search(self, key):
        temp = self.items.get(key, None)
        if temp:
            print('1', temp[1], temp[0])
        else:
            print('0')

    def min(self):
        if len(self.indexes) == 0:
            raise(Exception)

        key = self.indexes[0]
        print(key, '0', self.items[key][0])

    def max(self):
        if len(self.indexes) == 0:
            raise(Exception)

        i = int(len(self.indexes)/2)
        maxKey = self.indexes[i]

        index = i
        while i < len(self.indexes):
            if maxKey < self.indexes[i]:
                maxKey = self.indexes[i]
                index = i
            i += 1
        print(maxKey, index, self.items[maxKey][0])

    def extract(self):
        if len(self.indexes) == 0:
            raise(Exception)

        rootKey = self.indexes[0]
        rootData = self.items[rootKey][0]
        del self.items[rootKey]

        if len(self.indexes) > 1:
            self.indexes[0] = self.indexes.pop()

            # set top item index to 0
            self.items.update({self.indexes[0] : (self.items[self.indexes[0]][0], 0)})
            
            self.heapifyDown()
        else:
            self.indexes.pop()

        print(rootKey, rootData)

    def print(self):
        height = 0
        index = 0
        out = ''

        i = 0
        if len(self.indexes) == 0:
            out += '_\n'
            print('_')
            return

        while i < len(self.indexes):
            lineLen = 1 << height
            index += 1
            key = self.indexes[i]
            out += '[' + str(key) + ' ' + self.items[key][0]
            if height != 0:
                out += ' ' + str(self.indexes[self.parentIndex(i)])

            out += ']'

            if index == lineLen:
                out += '\n'
                index = 0
                height += 1
            else:
                out += ' '
            i += 1

        if index != 0 and index < lineLen:
            out += '_ ' * (lineLen - index)
            print(out[0:-1])
        else:
            print(out, end='')


cycle = True
heap = Heap()

while cycle:
    try:
        line = input()
        cmd = line.split(' ', 2)

        try:
            if len(cmd) == 1 and cmd[0] == '':
                continue

            if len(cmd) == 2 and cmd[0] == '' and cmd[1] == '':
                continue

            if cmd[0] == 'add':
                heap.add(int(cmd[1]), cmd[2])

            elif cmd[0] == 'set':
                heap.set(int(cmd[1]), cmd[2])

            elif cmd[0] == 'delete':
                heap.delete(int(cmd[1]))

            elif cmd[0] == 'search':
                heap.search(int(cmd[1]))

            elif cmd[0] == 'min':
                heap.min()

            elif cmd[0] == 'max':
                heap.max()

            elif cmd[0] == 'extract':
                heap.extract()

            elif cmd[0] == 'print':
                heap.print()

            else:
                raise(Exception)

        except Exception:
                print('error')
                continue

    except Exception:
        cycle = False

File: AlgorithmsAndDataStructures/mod2/Heap.cpp
#include <iostream>
#include <string>
#include <vector>
#include <unordered_map>
#include <exception>

class Heap
{
    std::unordered_map<int64_t, std::pair<std::string, unsigned int>> verts;
    std::vector<int64_t> keys;

protected:
    void swap(unsigned int &i, unsigned int &j)
    {
        auto key1 = keys[i];
        auto key2 = keys[j];

        keys[i] = key2;
        keys[j] = key1;

        std::swap(verts[key1].second, verts[key2].second);
    }

    bool hasParent(unsigned int &childIndex)
    {
        if (childIndex == 0)
            return false;
        if ((childIndex - 1) / 2 >= 0)
            return true;
        return false;
    }
    bool hasLChild(unsigned int &parentIndex)
    {
        if (parentIndex * 2 + 1 < keys.size())
            return true;
        return false;
    }
    bool hasRChild(unsigned int &parentIndex)
    {
        if (parentIndex * 2 + 2 < keys.size())
            return true;
        return false;
    }

    unsigned int getParentIndex(unsigned int &childIndex) { return (childIndex - 1) / 2; }
    unsigned int getLChildIndex(unsigned int &parentIndex) { return parentIndex * 2 + 1; }
    unsigned int getRChildIndex(unsigned int &parentIndex) { return parentIndex * 2 + 2; }

    void heapifyUp(unsigned int i)
    {
        if (!hasParent(i))
            return;

        unsigned int parentIndex = getParentIndex(i);
        while (hasParent(i) && keys[i] < keys[parentIndex])
        {
            swap(i, parentIndex);
            i = parentIndex;
            if (hasParent(i))
                parentIndex = getParentIndex(i);
        }
    }
    void heapifyDown(unsigned int i)
    {
        while (hasLChild(i))
        {
            auto smallerChildIndex = getLChildIndex(i);
            if (hasRChild(i) && keys[smallerChildIndex] > keys[getRChildIndex(i)])
                smallerChildIndex = getRChildIndex(i);

            if (keys[smallerChildIndex] > keys[i])
                break;
            swap(i, smallerChildIndex);
            i = smallerChildIndex;
        }
    }

public:
    Heap() = default;
    ~Heap() = default;

    void add(int64_t &key, std::string data)
    {
        if (verts.find(key) != verts.end())
            throw std::exception();

        verts.emplace(key, std::make_pair(data, keys.size()));
        keys.push_back(key);

        heapifyUp(keys.size() - 1);
    }

    void set(int64_t &key, std::string newData)
    {
        if (verts.find(key) == verts.end())
            throw std::exception();

        verts[key].first = newData;
    }

    void remove(int64_t &key)
    {
        if (verts.find(key) == verts.end())
            throw std::exception();

        unsigned int rmvIndex = verts[key].second;
        verts.erase(key);

        keys[rmvIndex] = keys[keys.size() - 1];
        keys.pop_back();

        if (keys.size() > rmvIndex)
        {
            verts[keys[rmvIndex]].second = rmvIndex;
            if (hasParent(rmvIndex) && keys[rmvIndex] < keys[getParentIndex(rmvIndex)])
                heapifyUp(rmvIndex);
            else if (hasLChild(rmvIndex))
                heapifyDown(rmvIndex);
        }
    }

    void search(int64_t &key)
    {
        if (verts.find(key) == verts.end())
            std::cout << '0' << std::endl;
        else
            std::cout << "1 " << verts[key].second << ' ' << verts[key].first << std::endl;
    }

    void min()
    {
        if (keys.size() == 0)
            throw std::exception();

        std::cout << keys[0] << ' ' << 0 << ' ' << verts[keys[0]].first << std::endl;
    }

    void max()
    {
        if (keys.size() == 0)
            throw std::exception();

        unsigned int index = keys.size() / 2;
        std::pair<int64_t, unsigned int> maxKey = {keys[index], index};
        for (int i = index; i < keys.size(); ++i)
        {
            if (maxKey.first < keys[i])
            {
                maxKey.first = keys[i];
                maxKey.second = i;
            }
        }

        std::cout << maxKey.first << ' ' << maxKey.second << ' ' << verts[maxKey.first].first << std::endl;
    }

    void extract()
    {
        if (keys.size() == 0)
            throw std::exception();

        std::cout << keys[0] << ' ' << verts[keys[0]].first << std::endl;

        verts.erase(keys[0]);
        if (keys.size() == 1)
            keys.pop_back();
        else
        {
            keys[0] = keys[keys.size() - 1];
            keys.pop_back();
            verts[keys[0]] = std::make_pair(verts[keys[0]].first, 0);

            heapifyDown(0);
        }
    }

    void print()
    {
        if (keys.empty())
        {
            std::cout << "_\n";
            return;
        }

        std::string output = "";
        auto index = 0;
        auto height = 0;
        for (unsigned int i = 0; i < keys.size(); ++i)
        {
            auto key = keys[i];
            ++index;
            if (i == 0)
                output += '[' + std::to_string(key) + ' ' + verts[key].first + ']';
            else
                output += '[' + std::to_string(key) + ' ' + verts[key].first + ' ' + std::to_string(keys[getParentIndex(i)]) + ']';

            if (index == 1 << height)
            {
                output += '\n';
                index = 0;
                ++height;
            }
            else
                output += ' ';
        }

        if (index != 0)
        {
            for (auto i = 0; i < (1 << height) - index; ++i)
                output += "_ ";
        }
        output.pop_back();
        std::cout << output << std::endl;
    }
};

int main()
{
    std::string line = "";
    Heap heap;
    while (std::cin >> line)
    {
        try
        {
            if (line == "")
                continue;

            if (line == "add")
            {
                int64_t key;
                std::cin >> key;
                std::string data;
                std::cin >> data;
                heap.add(key, data);
            }
            else if (line == "set")
            {
                int64_t key;
                std::cin >> key;
                std::string newData;
                std::cin >> newData;
                heap.set(key, newData);
            }
            else if (line == "delete")
            {
                int64_t key;
                std::cin >> key;
                heap.remove(key);
            }
            else if (line == "search")
            {
                int64_t key;
                std::cin >> key;
                heap.search(key);
            }
            else if (line == "min")
                heap.min();
            else if (line == "max")
                heap.max();
            else if (line == "extract")
                heap.extract();
            else if (line == "print")
                heap.print();
            else
            {
                std::cin.ignore(128, '\n');
                throw std::exception();
            }
        }
        catch (const std::exception &)
        {
            std::cout << "error\n";
        }
    }
    return 0;
}
